pkg/diagnostics/grpc_tracing.go

```diff
@@ -98,73 +101,89 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
-	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
-		if strings.Index(info.FullMethod, daprPackagePrefix) == 0 {
-			return handler(srv, ss)
-		}
-
-		var span trace.Span
-		spanName := info.FullMethod
+	return func(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		var (
+			span             trace.Span
+			spanKind         trace.SpanStartOption
+			prefixedMetadata map[string]string
+			reqSpanAttr      map[string]string
+			isProxied        bool
+		)
 
 		ctx := ss.Context()
-		md, _ := metadata.FromIncomingContext(ctx)
-
-		vals := md[GRPCProxyAppIDKey]
-		if len(vals) == 0 {
-			return errors.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
-		}
-
-		targetID := vals[0]
-		wrapped := grpcMiddleware.WrapServerStream(ss)
-		sc, _ := SpanContextFromIncomingGRPCMetadata(ctx)
 
-		var spanKind trace.SpanStartOption
-
-		if appID == targetID {
+		// This middleware is shared by multiple services and proxied requests, which need to be handled separately
+		switch {
+		// For gRPC service invocation, this generates ServerSpan
+		case strings.HasPrefix(info.FullMethod, daprInternalPrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindServer)
-		} else {
+
+		// For gRPC API, this generates ClientSpan
+		case strings.HasPrefix(info.FullMethod, daprRuntimePrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindClient)
+
+		// For proxied requests, this generates a span depending on whether this is the server (target) or client
+		default:
+			isProxied = true
+			md, _ := metadata.FromIncomingContext(ctx)
+			vals := md.Get(GRPCProxyAppIDKey)
+			if len(vals) == 0 {
+				return fmt.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
+			}
+			if appID == vals[0] {
+				spanKind = trace.WithSpanKind(trace.SpanKindServer)
+			} else {
+				spanKind = trace.WithSpanKind(trace.SpanKindClient)
+			}
 		}
 
+		// Overwrite context
+		sc, _ := SpanContextFromIncomingGRPCMetadata(ctx)
 		ctx = trace.ContextWithRemoteSpanContext(ctx, sc)
-		ctx, span = tracer.Start(ctx, spanName, spanKind)
-
+		ctx, span = tracer.Start(ctx, info.FullMethod, spanKind)
+		wrapped := grpcMiddleware.WrapServerStream(ss)
 		wrapped.WrappedContext = ctx
-		err := handler(srv, wrapped)
-
-		addSpanMetadataAndUpdateStatus(ctx, span, info.FullMethod, appID, nil, true)
 
-		UpdateSpanStatusFromGRPCError(span, err)
-		span.End()
-
-		return err
-	}
-}
+		isSampled := span.SpanContext().IsSampled()
+		if isSampled {
+			// users can add dapr- prefix if they want to see the header values in span attributes.
+			prefixedMetadata = userDefinedMetadata(ctx)
+			if isProxied {
+				reqSpanAttr = map[string]string{
+					daprAPISpanNameInternal: info.FullMethod,
+				}
+			} else {
+				reqSpanAttr = spanAttributesMapFromGRPC(appID, nil, info.FullMethod)
+			}
+		}
```

Comment:
Can this come after the `handler` is called? Combing this block and the `if` clause in line 164 together.

`addSpanMetadataAndUpdateStatus` function was called after `handler` was called in the previous code.

Reply:
@johnewart  commented the same above. In the unary interceptor there was a similar logic where `handler` is called sandwiched between two `if isSampled`. I think there's a chance that in the handler the context is changed and I do not feel comfortable changing that.